black_it/calibrator.py

```diff
@@ -133,32 +138,40 @@ def __init__(  # pylint: disable=too-many-arguments
             f"Selecting {self.n_jobs} processes for the parallel evaluation of the model"
         )
 
-        self.samplers_id_table = self._construct_samplers_id_table(samplers)
-
-    @property
-    def random_state(self) -> Optional[int]:
-        """Get the random state."""
-        return self._random_state
-
-    @random_state.setter
-    def random_state(self, random_state: Optional[int]) -> None:
-        """Set the random state."""
-        self._random_state = random_state
-        self._random_generator = default_rng(self.random_state)
+        self.scheduler = self.__validate_samplers_and_scheduler_constructor_args(
+            samplers, scheduler
+        )
 
-    @property
-    def random_generator(self) -> np.random.Generator:
-        """Get the random generator."""
-        return self._random_generator
+        self.samplers_id_table = self._construct_samplers_id_table(
+            list(self.scheduler.samplers)
+        )
 
-    def _get_random_seed(self) -> int:
-        """Get new random seed from the current random generator."""
-        return get_random_seed(self._random_generator)
+    @classmethod
+    def __validate_samplers_and_scheduler_constructor_args(
+        cls,
+        samplers: Optional[Sequence[BaseSampler]],
+        scheduler: Optional[BaseScheduler],
+    ) -> BaseScheduler:
+        """Validate the 'samplers' and the 'scheduler' arguments provided to the constructor."""
+        both_none = samplers is None and scheduler is None
+        both_not_none = samplers is not None and scheduler is not None
+        if both_none and both_not_none:
+            raise ValueError(
+                "only one between 'samplers' and 'scheduler' must be provided"
+            )
+
+        if samplers is not None:
+            return RoundRobinScheduler(samplers)
+
+        return cast(BaseScheduler, scheduler)
 
     def _set_samplers_seeds(self) -> None:
         """Set the calibration seed."""
-        for sampler in self.samplers:
-            sampler.random_state = self._get_random_seed()
+        self.scheduler.random_state = self.random_state
+
+        # "burn" seeds from the calibrator seed generator for backward compatibility
+        for _ in self.scheduler.samplers:
+            self._get_random_seed()
```

Comment:
this is really to show how the proposed change could, in theory, preserve the same behaviour.

Clearly "burning random seeds" is a bit odd thing to do.

Moreover, `self.scheduler.random_state = self.random_state` should be changed to `self.scheduler.random_state = self._get_random_seed()`, which necessarily detours to the old (exactly reproducible) behaviour, which is fine.